### ISSUE TITLE
Screenshot button implementation

### ISSUE THREAD

**Issue by zetok** [qTox/qTox#773]
Thanks to people in groupchat I got idea how to implement it UI-wise, here's quick mock-up made in gimp:
![mock-up](https://cloud.githubusercontent.com/assets/3148759/5061395/046380ca-6d8b-11e4-955c-aa86e4967c05.png)

Also, some explanation (sorry, I'm too lazy to write it again):
![explanation](https://cloud.githubusercontent.com/assets/3148759/5061397/1a074b6e-6d8b-11e4-9588-5839dd0ce872.png)

That "list" is supposed to be extensible in the future.

Hope it helps.

**Comment by tux3** [comment]
A screenshot button ? But, why ?

It's the OS's job to handles that.

**Comment by zetok** [comment]
@tux3 people seem to **really** want screenshot button in Tox client, and saying that it's OS's job doesn't help with that...

**Comment by ghost** [comment]
I remember that uTox has a screenshot button, just in place of qTox's attachment button.

**Comment by tux3** [comment]
I don't get it.

What's the usecase ? 
Why would it be better than hitting Print Screen ?
What's wrong with print screen ?

The GUI's gonna get cluttered and confusing if we had tons of buttons.

**Comment by zetok** [comment]
@tux3 to put it straight: use case is people who cannot into computers, and because of it they need screenshot in their qTox, because otherwise they won't be able to make a simple screenshot and send it.

Also, screenshot button isn't supposed to work like Print Screen, but offer a way to select only a part of screen…

**Comment by TheLastProject** [comment]
@tux3 This is exactly my opinion. To me, it makes absolutely no sense to have a screenshot button. The argument I've been given is that it's "easier", but I really don't feel like it is really an improvement. You can already either press PrintScreen and then paste it, or save the screenshot to a file and send the file through Tox. It feels quite redundant to reimplement a standard OS feature.

**Comment by tux3** [comment]
@zetok

If someone can't into computers, a simple GUI that does only what you'd expect will probably help more than tons of options imo.

> Also, screenshot button isn't supposed to work like Print Screen, but offer a way to select only a part of screen…

I would use a screenshot tool for that, not an IM client, don't you think ?
I don't think one of the main features of a Tox client should be a tool to take screenshots of a region of the screen. Not enough to justify a button anyway. A keyboard shortcut, maybe.

**Comment by dubslow** [comment]
It's one click + one drag to do it within qTox, whereas it's several clicks and interacting with a secondary program/the file system to do it otherwise (with the side effect of having another file on your hard drive). The integration is indeed quite convenient. I can arguably into computers and I would still appreciate this feature.

(I would also appreciate it if Cinammon had section selecting screenshots like Gnome 2+Compiz did, but hey. And I would still need to save it to file and send the file.)

Overall, if the code isn't terribly hard/huge, there's no reason to not have the feature. We can probably copy uTox's code for it as well. (Though, if that's specific to the OS/WM, as I suspect it is, then we'd have to figure out equivalent OS X code as well.) 

Additionally, desktop streaming is arguably rather more important than this, and if/once qTox gets that as well, this would be a "hole" in the feature list.

**Comment by tux3** [comment]
After a vote in presence of 239 people, ~10 of which voted, it was decided to implement this with a button that appears when hovering over the "Attach file" button, as mentioned in the OP.

http://strawpoll.me/3002675/r

**Comment by dubslow** [comment]
Where was this? lol

**Comment by tux3** [comment]
On #tox at 15:00 GMT :)

**Comment by ghost** [comment]
Why not just putting it into a (new) system menu using QActions? There's much place to insert some features like this and like others.
PS.: I don't like the blank screen shot button in uTox -- it even is unlabeled.

**Comment by zetok** [comment]
![some other opinions](https://cloud.githubusercontent.com/assets/3148759/5078801/65e5009a-6ea7-11e4-88da-2a6c824919fe.png)
![muh popularity](https://cloud.githubusercontent.com/assets/3148759/5078881/08200508-6ea8-11e4-87b4-18c12280aaec.png)

**Comment by zetok** [comment]
http://qt-project.org/doc/qt-4.8/desktop-screenshot.html
http://qt.developpez.com/doc/5.0-snapshot/desktop-screenshot/
for anyone willing to implement screenshot feature; that should make it ~easy, right?

Also, would be awesome if @ItsDuke could make a `screenshot` and `video capture` icons :3

---

EDIT: oh, and for people who still think that screenshot feature in qTox is bloat:

```
[15:36:05] <DoomBoom> qTox seems to have better possibilities for expansion thanks to the powerful Qt library and stuff
[15:37:01] <DoomBoom> really the 2 reasons I don't like it currently is it doesn't allow me to send screenpics easily or use my screen as video input
```

↑ words of someone who is implementing avatars in µTox; who could contribute his time to qTox…

Just bear in mind that it's not only attitude of developers…

**Comment by ItsDuke** [comment]
If someone could answer just a few questions about screenshots and desktop-sharing in regards to tox, that would be lovely. I think knowing these would better help me understand where these icons/buttons belong and how they'll be used.

1: How would the screenshot functionality work, and why should it be used instead of the system-default way?
2: Once a screenshot is taken, it is immediately sent as an attachment?
3: Does a user get control over the size of the screenshot taken?
4: What's the purpose of the video capture icon? You mean switching the source of video inputs from a webcam to a view of your screen?

**Comment by zetok** [comment]
\1. After pressing screenshot button, user would get something to select part (or whole) of screen, it would/should look somehow like that:
![screenshot button](https://cloud.githubusercontent.com/assets/3148759/5533294/e037311e-8a3c-11e4-8027-a01c861041dd.png)

As to why it's needed… **Users don't know about existence of system screenshot utility**. It is hidden, easily forgotten, etc.
Beside obvious user 'laziness' (why they just don't learn! :P ) often case is that one can't use system utility.
I for one wasn't able to quickly find screenshot utility on windows 8.1 in short moment I had displeasure to use it.

Even if there's no utility, theoretically one could press screeshot button on keyboard. _Theoretically_. In practice, there often is **no screenshot button on keyboard**, and even if there was one, user is again being told to `lrn2compute`.

Also, handy screenshot is nice thing to have to **quickly show** to your friend something from your screen.

\2. Great that you've asked. That question highlights what µTox does wrong with screenshot after it's taken: it immediately sends it. IMHO qTox shouldn't do that, but rather **after taking screenshot should show preview of it and ask user whether that's what user wants to send**.

\3. Yep.

\4. I worded it wrong, what I meant was `desktop sharing` instead of `video capture`. In same way as screenshot user would be able to decide how large part of screen should be streamed.

Not quite switching. 

I personally dislike how it's done in µTox, and I think that it should not switch from camera, but rather **provide another video for friend to see**. I.e., in case where one already is in video call, and wants to show a part of desktop using `desktop sharing`, friend would see video from camera, and separate video of part of shared desktop.

**Comment by dubslow** [comment]
As for 1, I can use the system utility, but it doesn't let me choose any section so I have to crop it with GIMP then find the file on my hdd to upload it, and altogether the process is easily 10x longer than if qTox offered a one-click one-drag manner to accomplish the same objective (and without cluttering my filesystem, though of course optionally enabling that should be allowed too).

**Comment by ItsDuke** [comment]
@zetok awesome, thanks a bunch for the explanation, it really helps to understand the functionality behind it to get a better idea of what the icons should look like.

As far as the desktop sharing goes, do you envision the desktop to be a separate single window, apart from the main chat window?

**Comment by srkunze** [comment]
I would love to see that feature in qTox as well. However, it looks like we are talking about two distinct features: Screenshots and Screensharing. Both should basically offer the same capabilities (entire screen, specific window, specific screen area).

**Comment by stqism** [comment]
a screenshot button sounds an awful lot like bloat for an OS feature.

**Comment by srkunze** [comment]
Well, at least screensharing with the three options would be great. Simple screenshots could be an easy addition later.

@stqism 
I know at first glance that looks like bloat because there are third-party apps for that. However, you also don't record your voice with a third-party app and then send it to your friend. The same goes for your webcam. All in one place serves a better UX; at least from my point of view.

**Comment by optimumtact** [comment]
That UI method of popping up buttons when you hover over a specific button is pretty terrible for user discovery.

They're only going to find out that feature exists if they happen to send a file.

**Comment by srkunze** [comment]
https://github.com/tux3/qTox/issues/1290

**Comment by AlexDaniel** [comment]
> That UI method of popping up buttons when you hover over a specific button is pretty terrible for user discovery.

^ that. Also it is a pain to use (you cannot go straight to the button you want). Just make a plain, regular button without this bullshit.

**Comment by zetok** [comment]
> > That UI method of popping up buttons when you hover over a specific button is pretty terrible for user discovery.
> 
> ^ that. Also it is a pain to use (you cannot go straight to the button you want). Just make a plain, regular button without this bullshit.

Perhaps could you show **how** to do that without ruining how qTox looks / feels, instead of only criticising?

If you've got better idea than mine, I'm all ears, but if you don't…

**Comment by AlexDaniel** [comment]
@zetok I already mentioned it. Just make these buttons always visible. That's it. The image you have provided looks OK to me (besides rounded corners on additional buttons).

"without ruining how qTox looks / feels"
It would feel like shit if you make buttons that pop out of nowhere and that are not accessible without hovering over something. There are just too many problems with the suggested dropleft solution. Want me to list them?

**Comment by TheLastProject** [comment]
@AlexDaniel That extra button in the text field looks absolutely hideous. We do need another solution than simply putting it there, really.

**Comment by AlexDaniel** [comment]
@TheLastProject well, making a button that pops out of another button is even more hideous. I'm just suggesting to use the lesser of two evils (which also happens to be easier to implement). If there is another solution that does not involve the text field - great.

**Comment by TheLastProject** [comment]
@AlexDaniel The lesser of two evils is permanently malforming/unbalancing the text field by having two buttons on the bottom and one on the top? I thoroughly disagree there.

**Comment by AlexDaniel** [comment]
@TheLastProject well, if we are so crazy about the design, then put smileys and "attach file" buttons on the top row. Now we have four buttons in a square, wooooow, how balanced this is! I am about to cum!

**Comment by TheLastProject** [comment]
@AlexDaniel You've succeeded in making clear you are not able to properly support your arguments and are not to be taken seriously. Thanks for making that clear, so we can properly discuss this with people who are worth the time discussing this with.

**Comment by AlexDaniel** [comment]
@TheLastProject I told you that I can make a list of arguments, if you wish. You didn't ask for it.
Also, look here:
![thinkpad x200 keyboard](http://files.progarm.org/2015-02-27-104406_905x616_scrot.png)
This looks so unbalanced. For example, look at these full-sized arrow keys! Nonsense!

**Comment by ProMcTagonist** [comment]
I think a user who wants to send something will investigate the send file button. The actions are highly related. 

We should think about what other things we want to send, and how they should be grouped. Voice clips are going to happen after ittner's inline media stuff is merged, and they're not file-system related.

Facebook has a tabs system on Android:

![android tabs](https://cloud.githubusercontent.com/assets/4358079/6410274/db7a079a-be22-11e4-81f0-3c36f8bf9e05.png)

Text is the default. When you click on another one, the text field and keyboard are replaced by the selector/creator for whatever you've chosen -- webcam, image chooser, stickers, voice recorder.

Their iOS implementation is rearranged but similar.

[Desktop Skype](https://support.skype.com/en/faq/FA3091/how-do-i-send-and-receive-files-using-skype-for-windows-desktop) appears to just stick everything inside a plus sign (in the call interface) or a paperclip (in the messaging interface).  

![call plus](https://az545221.vo.msecnd.net/skype-faq-media/faq_content/skype/screenshots/fa3091/fa3091_3.png)

![messaging paperclip](https://az545221.vo.msecnd.net/skype-faq-media/faq_content/skype/screenshots/fa3091/fa3091_2.png)

Their implementation is inconsistent and weird as-is, but may say something about possibly sticking it all in a popup from our paperclip. Paperclips mean Attach. That's generic and accepted. 

Just some things to think about.

**Comment by zetok** [comment]
@AlexDaniel to put it simply: unless you provide something **better**, buttons will be hidden.

**Comment by srkunze** [comment]
@ItsDuke 

> @zetok awesome, thanks a bunch for the explanation, it really helps to understand the functionality behind it to get a better idea of what the icons should look like.
> 
> As far as the desktop sharing goes, do you envision the desktop to be a separate single window, apart from the main chat window?

I cannot speak for @zetok but I would like to express my wish that it should be integrated into the main chat window (just as Skype do? - IIRC they have it above of the chat window).

Why? I cannot tell you why I dislike e.g. the peculiarity of GIMP of having 10s of windows floating around like hell. It just sucked. Thus I was very glad, that GIMP has the Single-Window-Mode now. I would like to have this for qTox as well from the start on.

**Comment by Doom032** [comment]
I also think it's better to not make it something that comes up by hovering. Best idea so far seems to be that clicking the paperclip would bring up a menu where you choose to send an attachment or make a screenshot, as @ProMcTagonist shows in Skype. Other approaches like Facebook's tab system might be interesting to look into, but would require a designer working it out a bit for tox.

Also @AlexDaniel could you be a bit more respectful? You can state your opinion without being so rude.

**Comment by tux3** [comment]
This has now been implemented by #1510.
